AdyenActions/Components/3DS2/Action handlers/3DS2+Delegated Authentication/ThreeDS2PlusDACoreActionHandler.swift

```diff
@@ -88,69 +83,228 @@
+                                       cardType: cardType,
+                                       cardNumber: cardNumber,
+                                       completion: completion)
+                } else {
+                    // setting the state to attempt the registration flow.
+                    delegatedAuthenticationState.attemptRegistration = true
+                    completion(.failure(.deviceIsNotRegistered))
                 }
-            } catch {
-                didFail(with: error, completionHandler: completionHandler)
             }
-            
         }
         
-        internal func performDelegatedAuthentication(_ fingerprintToken: ThreeDS2Component.FingerprintToken,
-                                                     completion: @escaping (Result<String, DelegateAuthenticationError>) -> Void) {
-            guard let delegatedAuthenticationInput = fingerprintToken.delegatedAuthenticationSDKInput else {
-                completion(.failure(DelegateAuthenticationError.authenticationFailed(cause: nil)))
-                return
+        // MARK: Delegated Authentication Approval
+        
+        private enum ApprovalFlowError: Error {
+            case fallbackTo3ds
+            case deviceIsNotRegistered
+            case authenticationServiceFailed(underlyingError: Error)
+            case removeCredentialServiceError(underlyingError: Error)
+        }
+        
+        private func showApprovalScreen(
+            delegatedAuthenticationInput: String,
+            cardType: CardType?,
+            cardNumber: String?,
+            completion: @escaping (Result<(daOutput: String, delete: Bool?), ApprovalFlowError>
+            ) -> Void
+        ) {
+            presenter.showApprovalScreen(
+                component: self,
+                cardDetails: (cardNumber, cardType),
+                approveAuthenticationHandler: { [weak self] in
+                    guard let self else { return }
+                    authenticate(delegatedAuthenticationInput: delegatedAuthenticationInput,
+                                 cardNumber: cardNumber,
+                                 authenticatedHandler: {
+                                     completion(.success((daOutput: $0, delete: nil)))
+                                 },
+                                 failedAuthenticationHandler: { [weak self] error in
+                                     guard let self else { return }
+                                     self.presenter.showAuthenticationError(component: self) {
+                                         completion(.failure(.authenticationServiceFailed(underlyingError: error)))
+                                     }
+                                 })
+                },
+                fallbackHandler: {
+                    completion(.failure(.fallbackTo3ds))
+                },
+                removeCredentialsHandler: { [weak self] in
```

Comment:
Same here (Could be that it's just Github formatting but maybe splitting would also improve readability)

Reply:
https://github.com/Adyen/adyen-ios/pull/1726#discussion_r1665801625